packages/compiler/src/core/binder.ts

```diff
@@ -36,38 +41,45 @@ import {
   TemplateParameterDeclarationNode,
   TypeSpecScriptNode,
   UnionStatementNode,
+  UnionVariantNode,
   UsingStatementNode,
 } from "./types.js";
 
 // Use a regular expression to define the prefix for TypeSpec-exposed functions
 // defined in JavaScript modules
 const DecoratorFunctionPattern = /^\$/;
-const SymbolTable = class extends Map<string, Sym> implements SymbolTable {
+// TODO: move to separate file
+const SymbolTable = class extends Map<string, Sym> implements MutableSymbolTable {
   duplicates = new Map<Sym, Set<Sym>>();
 
   constructor(source?: SymbolTable) {
     super();
     if (source) {
-      for (const [key, value] of source) {
-        // Note: shallow copy of value here so we can mutate flags on set.
-        super.set(key, { ...value });
-      }
-      for (const [key, value] of source.duplicates) {
-        this.duplicates.set(key, new Set(value));
-      }
+      this.include(source);
+    }
+  }
+
+  /** {@inheritdoc MutableSymboleTable} */
+  include(source: SymbolTable, parentSym?: Sym) {
+    for (const [key, value] of source) {
+      super.set(key, { ...value, parent: parentSym ?? value.parent });
+    }
+    for (const [key, value] of source.duplicates) {
+      this.duplicates.set(key, new Set(value));
     }
   }
 
   // First set for a given key wins, but record all duplicates for diagnostics.
   set(key: string, value: Sym) {
     const existing = super.get(key);
+
     if (existing === undefined) {
       super.set(key, value);
     } else {
+      // TODO: this was removed from the binder explicitly, figure out why?
```

Comment:
@bterlson in your pr the check was removed below does this seems like it was expected, without that can't track ambiguous symbols